### ISSUE TITLE
Aligning in speaker

### ISSUE THREAD

**Issue by zwx8981** [airsplay/R2R-EnvDrop#5]
https://github.com/airsplay/R2R-EnvDrop/blob/4c115853b6e53dd245f965e99d63579372d7ebdb/r2r_src/speaker.py#L243

Hi, I am confused about the aligning operation (by your comments). It seems that you ignore the last element of the predicted logits (the logits of 'EOS' or 'PAD') and the 'BOS' of the target when you compute the loss during training the speaker (which, in my view, make the logits and target unalgined...).  Can you explain how does this operation make the logits and target aligned?

**Comment by airsplay** [comment]
Hmmm. 

Let me explain it with an example. Suppose the sequence is "hello world", the input and desired output in teacher_forcing would be:
```
Time:                  1       2       3       4
Input:              <bos>    hello   world   <eos>  
Desired Output:     hello    world   <eos>    ???
```
And the logit outputed by the model is:
```
Time:                  1       2       3       4
Input:              <bos>    hello   world   <eos>  
Output Logit:        L_0      L_1     L_2     L_3 
```

Thus the aligned logit and target would be
```
logit[:-1]:            L_0     L_1     L_2  
Input[1:]:            hello   world   <eos>  
```
And it is what the code does. 

Hope this answers your question!

**Comment by zwx8981** [comment]
Oh, that's very clear! Thank you so much!
